### PR TITLE
drivers: sensor: stm32_temp

### PR DESCRIPTION
**Pull request by Delsian**
STM32 internal temperature sensor driver.
This sensor can be used to measure the temperature of the CPU
and its surroundings.
Related issue: https://github.com/zephyrproject-rtos/zephyr/issues/34538

Signed-off-by: Eug Krashtan <eug.krashtan@gmail.com>

### CODE REVIEW

**Comment by Delsian** [comment]
I'd like to make a change in sensor.h after merging https://github.com/zephyrproject-rtos/zephyr/pull/34679
Please review it first.

**Review by MaureenHelm — CHANGES_REQUESTED** [review]
Thanks for your contribution!

Please split the changes into separate commits:
- drivers: adc: Define max acquisition time
- drivers: sensor: Add helper function to convert double to struct sensor_value
- drivers: sensor: Introduce stm32 die temperature driver
- samples: sensor: Add stm32_temp sample application

Add `samples/sensor/stm32_temp_sensor/sample.yaml` to ensure the sample gets built in CI.

Add this sensor to `tests/drivers/build_all/sensors_i_z.conf`

**Comment by Delsian** [comment]
@MaureenHelm Thank you. I did all the requested changes.

File: drivers/adc/adc_stm32.c
Comment (erwango) [inline drivers/adc/adc_stm32.c:491]: Given the code below, I wonder in which conditions you can get this 'unused variable' warning,
can you clarify ?

**Review by erwango — CHANGES_REQUESTED** [review]
Few changes required, overall looks really good, 
Thanks for this work!

**Comment by erwango** [comment]
^^ @StefJar

**Comment by StefJar** [comment]
thx @Delsian for that pull. @MaureenHelm guess since 2 years there is talking about the stm32 adc. As I am not an expert of all stm32 platforms, I can only check that commit against the stm32F412.

some issues for the stm32F412

https://github.com/zephyrproject-rtos/zephyr/blob/6857eb198eb937b8d30c17203e5c46a54d67537b/drivers/adc/adc_stm32.c#L712 this is not activating the internal VRef channel at the F4 series (but needs to be done)

https://github.com/zephyrproject-rtos/zephyr/blob/6857eb198eb937b8d30c17203e5c46a54d67537b/drivers/adc/adc_stm32.c#L331maps the an index to a external channel. When checking the LL for the stm32F412 the channel 17 values for the internal and external channel are different. Means that with this code you are not able to select the internal channel.

@MaureenHelm IMHO think it is might worth full to think about an adc channel not as an number. Instead as a tuple. Like
adc_ch1 = (<param1> <param2> ... <param N>)
for the stm32 adc this goes like adc_ch17 -> adc_chVRef = (<17> <internal>) and  adc_ch17 = (<17> <external>)
each adc implementation can use its own channel pattern via dts

File: drivers/adc/adc_stm32.c
Comment (Delsian) [reply drivers/adc/adc_stm32.c:491]: @erwango For example, modules\hal\stm32\stm32cube\stm32f0xx\drivers\include\stm32f0xx_ll_adc.h:
```
#define __LL_ADC_COMMON_INSTANCE(__ADCx__)                                     \
  (ADC1_COMMON)
```

**Comment by Delsian** [comment]
@StefJar Thank you for comment. Just checked with STM32F412RE. 
Bit TSVREFE sets ok, `adc_stm32_set_common_path(dev)` was resolved to
```
__STATIC_INLINE void LL_ADC_SetCommonPathInternalCh(ADC_Common_TypeDef *ADCxy_COMMON, uint32_t PathInternal)
{
  MODIFY_REG(ADCxy_COMMON->CCR, ADC_CCR_TSVREFE | ADC_CCR_VBATE, PathInternal);
}
```
in file stm32f4xx_ll_adc.h and debugger shows CCR=0x00800000 (i.e.  ADC_CCR_TSVREFE set)
So this part of the code has no problem in STM32F4
Values from ADC also changed as expexted when heating, but STM32F4 has another formula to calculate the temperature.
Will change `stm32_temp_channel_get` accordingly.

File: drivers/sensor/stm32_temp/stm32_temp.c
Comment (Delsian) [reply drivers/sensor/stm32_temp/stm32_temp.c]: @erwango Checked with F1 and F4 series. Unfortunately, I have no other device series.

**Comment by Delsian** [comment]
@erwango @StefJar Probably have the sense to move AVGSLOPE and V25 values to Kconfig section? And unify formula for all chips? Will investigate this ASAP

File: drivers/sensor/stm32_temp/stm32_temp.c
Comment (erwango) [inline drivers/sensor/stm32_temp/stm32_temp.c:50]: If V25 and AVGSLOPE are likely to change between each soc, can we consider getting those from device tree ?

Comment (Delsian) [reply drivers/sensor/stm32_temp/stm32_temp.c:50]: @erwango Exactly!

Comment (StefJar) [reply drivers/sensor/stm32_temp/stm32_temp.c:50]: @Delsian I checked the driver file. For me it looks that the temperature enabling is not activated  (see https://github.com/zephyrproject-rtos/zephyr/blob/4c2fe4b39c0a29a535fd0fc21a706a53e53a2975/drivers/adc/adc_stm32.c#L697)

What I was mainly referring was the use of the `__LL_ADC_DECIMAL_NB_TO_CHANNEL(index)` in combination with the dts binding.

form stm32F4 HAL
```
_LL_ADC_CHANNEL_TO_DECIMAL_NB
Description:
•
Helper macro to get ADC channel number in decimal format from literals LL_ADC_CHANNEL_x.
Parameters:
•
__CHANNEL__: This parameter can be one of the following values:
–
LL_ADC_CHANNEL_0
–
LL_ADC_CHANNEL_1
–
LL_ADC_CHANNEL_2
–
LL_ADC_CHANNEL_3
–
LL_ADC_CHANNEL_4
–
LL_ADC_CHANNEL_5
–
LL_ADC_CHANNEL_6
–
LL_ADC_CHANNEL_7
–
LL_ADC_CHANNEL_8
–
LL_ADC_CHANNEL_9
–
LL_ADC_CHANNEL_10
–
LL_ADC_CHANNEL_11
–
LL_ADC_CHANNEL_12
–
LL_ADC_CHANNEL_13
–
LL_ADC_CHANNEL_14
–
LL_ADC_CHANNEL_15
–
LL_ADC_CHANNEL_16
–
LL_ADC_CHANNEL_17
–
LL_ADC_CHANNEL_18
–
LL_ADC_CHANNEL_VREFINT (1)
–
LL_ADC_CHANNEL_TEMPSENSOR (1)(2)
–
LL_ADC_CHANNEL_VBAT (1)
Return value:
•
Value: between Min_Data=0 and Max_Data=18
Notes:
•
Example: __LL_ADC_CHANNEL_TO_DECIMAL_NB(LL_ADC_CHANNEL_4) will return decimal number
"4". The input can be a value from functions where a channel number is returned, either defined with
number or with bitfield (only one bit must be set).
```
As you can see at the driver

https://github.com/zephyrproject-rtos/zephyr/blob/6857eb198eb937b8d30c17203e5c46a54d67537b/drivers/adc/adc_stm32.c#L527

this line excludes the VREF, TEMPSENSOR and VBAT

check the define (STM32_CHANNEL_COUNT is set to 20)

I hope that you/we may alter the driver so we can use VREF etc. (What is my goal)

Comment (Delsian) [reply drivers/sensor/stm32_temp/stm32_temp.c]: Ok, I moved all parameters to DTS overlay, and now both existing boards (F103 and F412) work fine.
Can somebody check with other series and propose overlays with correct values?

Comment (Delsian) [reply drivers/sensor/stm32_temp/stm32_temp.c]: > @Delsian I checked the driver file. For me it looks that the temperature enabling is not activated for the stm32F4 series (see
> 
> https://github.com/zephyrproject-rtos/zephyr/blob/4c2fe4b39c0a29a535fd0fc21a706a53e53a2975/drivers/adc/adc_stm32.c#L697
> 
> )

@StefJar I suppose we can drop this line as it not used anymore. Now TSVREFE init occurs in channel initialization.

**Comment by Delsian** [comment]
@erwango One small question - is it possible to use something like "**platform_allow**" in sample.yaml, but select the whole STM32 family instead of a board name like nucleo_xxx?
Or how to restrict buildkite in a proper way?

**Comment by erwango** [comment]
> is it possible to use something like "platform_allow" in sample.yaml, but select the whole STM32 family instead of a board name like nucleo_xxx?
Or how to restrict buildkite in a proper way?

You can use 'filter' field, for instance:
```
tests:
  sample.drivers.flash.shell:
    tags: flash shell
    filter: CONFIG_FLASH_HAS_DRIVER_ENABLED
```
You can also base the filter on dts, which could be even more fine-grained:
```
    filter: dt_compat_enabled("st,stm32-window-watchdog")
```
You can also use operators: `not`, `and`, `or`, ...

**Comment by Delsian** [comment]
@erwango Thank you. I've tried the second variant without success, but finally, I understand where is my problem. Hope It's the final push (fingers crossed)

File: drivers/adc/adc_stm32.c
Comment (FRASTM) [inline drivers/adc/adc_stm32.c:466]: maybe a slight comment  could explain why '7', where the value comes from

**Review by FRASTM — COMMENTED** [review]
Do you have a result of the sample temp application running ?

File: drivers/adc/adc_stm32.c
Comment (Delsian) [reply drivers/adc/adc_stm32.c:466]: @FRASTM 7 is maximal index (and represents longest time, as required in temperature conversion) in
```
static const uint16_t acq_time_tbl[8] = {2, 8, 14, 29, 42, 56, 72, 240};
```
Have no idea why `adc_stm32.c` uses magic numbers instead definitions.

**Comment by Delsian** [comment]
@FRASTM Yes, right now it runs here on devboard since yesterday. Here is short heating of chip body with soldering iron:
```
Current temperature: 24.2 °C
Current temperature: 24.6 °C
Current temperature: 25.4 °C
Current temperature: 25.0 °C
Current temperature: 25.4 °C
Current temperature: 26.2 °C
Current temperature: 26.2 °C
Current temperature: 26.2 °C
Current temperature: 27.0 °C
Current temperature: 27.0 °C
Current temperature: 27.8 °C
Current temperature: 27.8 °C
Current temperature: 28.2 °C
Current temperature: 28.2 °C
Current temperature: 28.6 °C
Current temperature: 29.4 °C
Current temperature: 29.0 °C
Current temperature: 29.8 °C
Current temperature: 30.6 °C
Current temperature: 29.8 °C
Current temperature: 31.4 °C
Current temperature: 31.4 °C
Current temperature: 31.4 °C
Current temperature: 31.4 °C
Current temperature: 31.8 °C
Current temperature: 32.6 °C
Current temperature: 31.0 °C
Current temperature: 29.8 °C
Current temperature: 29.0 °C
Current temperature: 28.2 °C
Current temperature: 28.2 °C
Current temperature: 27.4 °C
Current temperature: 27.0 °C
Current temperature: 27.4 °C
Current temperature: 26.6 °C
Current temperature: 26.2 °C
Current temperature: 26.2 °C
Current temperature: 26.2 °C
Current temperature: 25.8 °C
Current temperature: 25.8 °C
Current temperature: 25.4 °C
```

**Review by FRASTM — APPROVED** [review]

File: samples/sensor/stm32_temp_sensor/sample.yaml
Comment (erwango) [inline samples/sensor/stm32_temp_sensor/sample.yaml:6]: Don't use `platform_allow` with `filter`.

File: samples/sensor/stm32_temp_sensor/boards/nucleo_f401re.overlay
Comment (erwango) [inline samples/sensor/stm32_temp_sensor/boards/nucleo_f401re.overlay:20]: adc1 is already part of nucleo_f401re default configuration (`nucleo_f401re.dts`)
.overlay is added on top of this default configuration, so this is redundant.

File: samples/sensor/stm32_temp_sensor/sample.yaml
Comment (erwango) [inline samples/sensor/stm32_temp_sensor/sample.yaml:10]: I think the reasoning to define filter should be:
`What is the prerequisite for a board to be compatible with this sample ?`

On my side, I'll try `dt_compat_enabled_with_label("st,stm32-adc", "ADC_1")`.
I haven't tested, so please have a try.
If not working, let's forget about the filter and stick to `platform_allow`

**Review by erwango — CHANGES_REQUESTED** [review]

File: samples/sensor/stm32_temp_sensor/sample.yaml
Comment (Delsian) [reply samples/sensor/stm32_temp_sensor/sample.yaml]: @erwango I'm not sure about proposed filter. It should build a sample with any STM32 device with ADC enabled.
Maybe, there is any solution to add .overlay when CONFIG_STM32_TEMP enabled? And this allows all builds with CONFIG_SOC_FAMILY_STM32=y

File: samples/sensor/stm32_temp_sensor/sample.yaml
Comment (erwango) [inline samples/sensor/stm32_temp_sensor/sample.yaml:9]: For some reason this is not producing expected result.
In order to avoid blocking this to get to DV2.6.0, let's go with:
```
    depends_on: adc
    tags: sensors
    build_only: true
    platform_allow: nucleo_f401re stm32f103_mini
```
We'll try to find a better approach later.

**Review by erwango — CHANGES_REQUESTED** [review]

File: samples/sensor/stm32_temp_sensor/sample.yaml
Comment (erwango) [reply samples/sensor/stm32_temp_sensor/sample.yaml:9]: Note, you can see effect on twister with :
```
twister -T samples/sensor/stm32_temp_sensor/
```

Comment (Delsian) [reply samples/sensor/stm32_temp_sensor/sample.yaml]: Thanks, done!

**Review by erwango — APPROVED** [review]

**Comment by erwango** [comment]
@MaureenHelm can you have a new look ?

**Comment by Delsian** [comment]
@FRASTM  I have question about STM32G4 HAL. In this HAL there is no definition for ADC_CHANNEL_TEMPSENSOR, common for all other HALs. Is it possible to add 
```
#define ADC_CHANNEL_TEMPSENSOR ADC_CHANNEL_TEMPSENSOR_ADC1
```
to file `stm32cube\stm32g4xx\drivers\include\stm32g4xx_hal_adc.h` ?

**Review by ABOSTM — APPROVED** [review]
Minor remarkl, otherwise LGTM

**Comment by Delsian** [comment]
@ABOSTM Yes, fixed with ARRAY_SIZE() instead of sizeof(), as this table contains uint16 values.

**Comment by erwango** [comment]
> @erwango Sorry for disturbing, but looks like I need your help again. Probably my change in file `tests/drivers/build_all/sensors_i_z.conf` crashes buildkite for some reason, but I have no idea why.
> Can you take a short look, please?

Seems like a buildkite limitation, but that's ok.
What we miss is @MaureenHelm to revisit and approve.

**Comment by galak** [comment]
> @erwango Sorry for disturbing, but looks like I need your help again. Probably my change in file `tests/drivers/build_all/sensors_i_z.conf` crashes buildkite for some reason, but I have no idea why.
> Can you take a short look, please?

Don't worry about that, will look over the CI issue once we have @MaureenHelm's approval.

**Review by MaureenHelm — CHANGES_REQUESTED** [review]

File: drivers/sensor/stm32_temp/stm32_temp.c
Comment (Delsian) [reply drivers/sensor/stm32_temp/stm32_temp.c]: @MaureenHelm Done

**Review by MaureenHelm — APPROVED** [review]

**Pull request merged**
